File: src/mantle/blocks/abstracts/InventorySlab.scala
package mantle.blocks.abstracts

import net.minecraft.block.material.Material
import net.minecraft.entity.Entity
import net.minecraft.util.AxisAlignedBB
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import java.util.List

/**
 * Slab version of InventoryBlock.
 *
 * @author mDiyo
 * @author Sunstrike
 */
abstract class InventorySlab(id:Int, mat:Material) extends InventoryBlock(id, mat) {

  override def renderAsNormalBlock: Boolean = false

  override def isOpaqueCube: Boolean = false

  override def shouldSideBeRendered(world: IBlockAccess, x: Int, y: Int, z: Int, side: Int): Boolean = {
    if (side > 1) return super.shouldSideBeRendered(world, x, y, z, side)
    val meta: Int = world.getBlockMetadata(x, y, z)
    val top: Boolean = (meta | 8) == 1
    if ((top && side == 0) || (!top && side == 1)) return true
    super.shouldSideBeRendered(world, x, y, z, side)
  }

  override def addCollisionBoxesToList(world: World, x: Int, y: Int, z: Int, axisalignedbb: AxisAlignedBB, arraylist: List[_], entity: Entity) {
    setBlockBoundsBasedOnState(world, x, y, z)
    super.addCollisionBoxesToList(world, x, y, z, axisalignedbb, arraylist, entity)
  }

  override def setBlockBoundsForItemRender() {
    setBlockBounds(0.0F, 0.0F, 0.0F, 1.0F, 0.5F, 1.0F)
  }

  override def setBlockBoundsBasedOnState(world: IBlockAccess, x: Int, y: Int, z: Int) {
    val meta: Int = world.getBlockMetadata(x, y, z) / 8
    val minY: Float = if (meta == 1) 0.5F else 0.0F
    val maxY: Float = if (meta == 1) 1.0F else 0.5F
    setBlockBounds(0.0F, minY, 0F, 1.0F, maxY, 1.0F)
  }

  override def onBlockPlaced(par1World: World, blockX: Int, blockY: Int, blockZ: Int, side: Int, clickX: Float, clickY: Float, clickZ: Float, metadata: Int): Int = {
    if (side == 1) return metadata
    if (side == 0 || clickY >= 0.5F) return metadata | 8
    metadata
  }

  override def damageDropped(meta: Int): Int = meta % 8

}

File: src/mantle/blocks/abstracts/InventorySlab.java
package mantle.blocks.abstracts;

import java.util.List;

import net.minecraft.block.material.Material;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.Entity;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.BlockPos;
import net.minecraft.util.EnumFacing;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;

/**
 * Slab version of InventoryBlock.
 *
 * @author mDiyo
 */
public abstract class InventorySlab extends InventoryBlock
{
    public InventorySlab(Material material)
    {
        super(material);
    }

    /* Rendering */
    @Override
    public boolean renderAsNormalBlock ()
    {
        return false;
    }

    @Override
    public boolean isOpaqueCube ()
    {
        return false;
    }

    @Override
    public boolean shouldSideBeRendered (IBlockAccess world, BlockPos pos, EnumFacing side)
    {
        if (side > 1)
            return super.shouldSideBeRendered(world, pos, side);

        int meta = world.getBlockMetadata(pos);
        boolean top = (meta | 8) == 1;
        if ((top && side == 0) || (!top && side == 1))
            return true;

        return super.shouldSideBeRendered(world, pos, side);
    }

    @Override
    @SuppressWarnings("rawtypes")
    public void addCollisionBoxesToList (World world, BlockPos pos, IBlockState state, AxisAlignedBB axisalignedbb, List arraylist, Entity entity)
    {
        setBlockBoundsBasedOnState(world, pos);
        super.addCollisionBoxesToList(world,pos ,state,  axisalignedbb, arraylist, entity);
    }

    public void setBlockBoundsForItemRender ()
    {
        setBlockBounds(0.0F, 0.0F, 0.0F, 1.0F, 0.5F, 1.0F);
    }

    public void setBlockBoundsBasedOnState (IBlockAccess world, BlockPos pos, IBlockState state)
    {
        int meta = world.getBlockState(pos) / 8;
        float minY = meta == 1 ? 0.5F : 0.0F;
        float maxY = meta == 1 ? 1.0F : 0.5F;
        setBlockBounds(0.0F, minY, 0F, 1.0F, maxY, 1.0F);
    }

    public int onBlockPlaced (World par1World, int blockX, int blockY, int blockZ, int side, float clickX, float clickY, float clickZ, int metadata)
    {
        if (side == 1)
            return metadata;
        if (side == 0 || clickY >= 0.5F)
            return metadata | 8;

        return metadata;
    }

    public int damageDropped (int meta)
    {
        return meta % 8;
    }
}